File: 2025-01-22/Median_Of_Two_Sorted_Arrays.java
package practice;

public class Median_Of_Two_Sorted_Arrays {
    public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums2.length < nums1.length) return findMedianSortedArrays(nums2, nums1);

        int low1 = 0, high1 = nums1.length;

        while(low1 <= high1) {
            int mid1 = (high1 + low1) / 2;
            int mid2 = (nums1.length + nums2.length + 1) / 2 - mid1;

            int l1 = Integer.MIN_VALUE, r1 = Integer.MAX_VALUE;
            int l2 = Integer.MIN_VALUE, r2 = Integer.MAX_VALUE;
            if(mid1>0) l1 = nums1[mid1 - 1];
            if(mid1<nums1.length) r1 = nums1[mid1];
            if(mid2>0) l2 = nums2[mid2 - 1];
            if(mid2<nums2.length) r2 = nums2[mid2];

            if(l1 <= r2 && l2 <= r1) {
                if((nums1.length + nums2.length) % 2 == 0)
                    return (Math.max(l1, l2) + Math.min(r1, r2)) / 2.0;
                return Math.max(l1, l2);
            }
            else if(l1 > r2) {
                high1 = mid1 - 1;
            }
            else {
                low1 = mid1 + 1;
            }
        }

        return 0.0;
    }
    public static void main(String[] args) {
        int[] arr = {3,4};
        int[] arr2 = {1,2};

        System.out.println(findMedianSortedArrays(arr, arr2));
    }
}

File: 2026-02-10/Median_Of_Two_Sorted_Arrays.py
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1, n2 = len(nums1), len(nums2)
        h = (n1 + n2) // 2
        if n1 > n2:
            nums1, nums2 = nums2, nums1
            n1, n2 = n2, n1

        l, r = 0, n1-1
        while l <= r:
            i = l + (r - l) // 2
            j = h - i - 2

            l1 = nums1[i] if i >= 0 else float('-inf')
            r1 = nums1[i+1] if i+1 < n1 else float('inf')
            l2 = nums2[j] if j >= 0 else float('-inf')
            r2 = nums2[j+1] if j + 1 < n2 else float('inf')

            if l1 <= r2 and l2 <= r1:
                if (n1 + n2) & 1:
                    return min(r1,r2)
                else:
                    return (max(l1, l2) + min(r1,r2)) / 2.0
            elif l1 > r2:
                r = i - 1
            else:
                l = i + 1

        return nums1[l]





sol = Solution()
print(sol.findMedianSortedArrays([1,3,4,10], [6,7,12,14,18,20,21]))
